File: web/src/lib/bins.ts
import Supercluster from "supercluster";
import raw from "../../data/bins.json";

export type BinKind = "recycling" | "ewaste";

interface RawBin {
  n: string;
  p: string;
  t: string;
  s: string[];
  y: number;
  x: number;
}

export interface Bin {
  id: number;
  name: string;
  postal: string;
  kind: BinKind;
  streams: string[];
  lat: number;
  lng: number;
}

export const BINS: Bin[] = (raw as RawBin[]).map((r, i) => ({
  id: i,
  name: r.n,
  postal: r.p,
  kind: r.t as BinKind,
  streams: r.s,
  lat: r.y,
  lng: r.x,
}));

type Props = { id: number; kind: BinKind };

/**
 * 13,000 points must never reach a phone as one payload. The index lives on the
 * server; the client only ever receives what fits its current viewport.
 */
const indexes = new Map<string, Supercluster<Props>>();

function indexFor(kinds: BinKind[]): Supercluster<Props> {
  const key = [...kinds].sort().join(",");
  const cached = indexes.get(key);
  if (cached) return cached;

  const idx = new Supercluster<Props>({ radius: 70, maxZoom: 17, minPoints: 3 });
  idx.load(
    BINS.filter((b) => kinds.includes(b.kind)).map((b) => ({
      type: "Feature" as const,
      properties: { id: b.id, kind: b.kind },
      geometry: { type: "Point" as const, coordinates: [b.lng, b.lat] },
    })),
  );
  indexes.set(key, idx);
  return idx;
}

export interface ClusterPoint {
  cluster: true;
  id: number;
  count: number;
  lat: number;
  lng: number;
}

export interface SinglePoint {
  cluster: false;
  bin: Bin;
  lat: number;
  lng: number;
}

export type MapPoint = ClusterPoint | SinglePoint;

export function pointsInView(
  bbox: [number, number, number, number],
  zoom: number,
  kinds: BinKind[],
): MapPoint[] {
  return indexFor(kinds)
    .getClusters(bbox, Math.round(zoom))
    .map((f) => {
      const [lng, lat] = f.geometry.coordinates;
      if (f.properties && "cluster" in f.properties && f.properties.cluster) {
        return {
          cluster: true as const,
          id: f.properties.cluster_id as number,
          count: f.properties.point_count as number,
          lat,
          lng,
        };
      }
      return { cluster: false as const, bin: BINS[(f.properties as Props).id], lat, lng };
    });
}

/** Haversine, in metres. */
export function metresBetween(aLat: number, aLng: number, bLat: number, bLng: number): number {
  const R = 6_371_000;
  const dLat = ((bLat - aLat) * Math.PI) / 180;
  const dLng = ((bLng - aLng) * Math.PI) / 180;
  const p =
    Math.sin(dLat / 2) ** 2 +
    Math.cos((aLat * Math.PI) / 180) * Math.cos((bLat * Math.PI) / 180) * Math.sin(dLng / 2) ** 2;
  return Math.round(2 * R * Math.asin(Math.sqrt(p)));
}

export function nearest(lat: number, lng: number, kinds: BinKind[], limit: number) {
  return BINS.filter((b) => kinds.includes(b.kind))
    .map((b) => ({ ...b, metres: metresBetween(lat, lng, b.lat, b.lng) }))
    .sort((a, b) => a.metres - b.metres)
    .slice(0, limit);
}

File: fastAPI_chatbot/app/tools/bins.py
"""Nearest-bin lookup over the shared NEA dataset.

13,004 real collection points fetched from data.gov.sg by `scripts/fetch-bins.py`
at the repo root — 12,291 blue recycling bins and 713 e-waste points. The file is
READ from web/data/bins.json, never copied: two copies drift, and the one you demo
will be the stale one.

No spatial index. A brute-force haversine over 13k rows is a few milliseconds, and
an index would be one more thing to keep correct for no measurable gain. No Google
Places, no OneMap, no key, and it works with the wifi off.
"""

import json
import logging
import math
from dataclasses import dataclass
from functools import lru_cache
from pathlib import Path
from typing import Literal

from app.config import settings

logger = logging.getLogger(__name__)

BinKind = Literal["recycling", "ewaste"]

# Opens the phone's native maps app with walking directions from wherever the user
# is standing. A real directions API would return a polyline we have nowhere to draw.
DIRECTIONS_URL = "https://www.google.com/maps/dir/?api=1&destination={lat},{lng}"


@dataclass(frozen=True)
class Bin:
    name: str
    postal: str
    kind: BinKind
    streams: tuple[str, ...]
    lat: float
    lng: float


@dataclass(frozen=True)
class NearbyBin:
    bin: Bin
    metres: int

    @property
    def directions_url(self) -> str:
        return DIRECTIONS_URL.format(lat=self.bin.lat, lng=self.bin.lng)


@lru_cache(maxsize=1)
def _bins() -> tuple[Bin, ...]:
    """The dataset, loaded once per process.

    A missing file is logged and returns nothing rather than raising: the chatbot
    still answers disposal questions without it, and losing the bin finder should
    not take down /chat.
    """
    path = Path(settings.BINS_PATH)
    if not path.exists():
        logger.error("Bin dataset not found at %s - location answers disabled.", path)
        return ()

    try:
        raw = json.loads(path.read_text(encoding="utf-8"))
    except (json.JSONDecodeError, OSError):
        logger.exception("Could not read %s - location answers disabled.", path)
        return ()

    # Compact keys, as written by scripts/build-bins.py: n=name, p=postal, t=type,
    # s=streams, y=lat, x=lng.
    loaded = tuple(
        Bin(
            name=r["n"],
            postal=r["p"],
            kind=r["t"],
            streams=tuple(r.get("s", ())),
            lat=r["y"],
            lng=r["x"],
        )
        for r in raw
    )
    logger.info("Loaded %d bins from %s", len(loaded), path)
    return loaded


def metres_between(a_lat: float, a_lng: float, b_lat: float, b_lng: float) -> int:
    """Haversine, in metres. Mirrors metresBetween() in web/src/lib/bins.ts."""
    radius = 6_371_000
    d_lat = math.radians(b_lat - a_lat)
    d_lng = math.radians(b_lng - a_lng)
    h = (
        math.sin(d_lat / 2) ** 2
        + math.cos(math.radians(a_lat))
        * math.cos(math.radians(b_lat))
        * math.sin(d_lng / 2) ** 2
    )
    return round(2 * radius * math.asin(math.sqrt(h)))


def find_bins(
    lat: float, lng: float, kind: BinKind | None = None, limit: int = 3
) -> list[NearbyBin]:
    """The nearest bins of a kind, closest first.

    `kind` of None searches both, which is right for a vague "where do I recycle
    this" where the item was never identified.
    """
    candidates = [b for b in _bins() if kind is None or b.kind == kind]
    ranked = sorted(
        (NearbyBin(b, metres_between(lat, lng, b.lat, b.lng)) for b in candidates),
        key=lambda n: n.metres,
    )
    return ranked[:limit]


def all_bins() -> tuple[Bin, ...]:
    """Every loaded bin. Used by the geocoder to resolve postal codes offline."""
    return _bins()


def count() -> int:
    """How many bins loaded. Surfaced on /health to diagnose a bad BINS_PATH."""
    return len(_bins())


def reset() -> None:
    """Drop the cached dataset. For tests."""
    _bins.cache_clear()
